### DISCUSSION TITLE
Why are opposite relations required?

### DISCUSSION THREAD

**Discussion by ErikDakoda** [prisma/prisma#21380] [Q&A]
### Question

Why are opposite relations required? This requirement results in User becoming a Frankenstein's monster with a reference to every other model in the codebase. In the real world, I will never use `prisma.user.findUnique()` with all of its related Posts. Instead, I will use `prisma.post.findMany()` specifying a userId in the where clause. I mean, it's nice to have in certain circumstances, but why is it a requirement? It's certainly not a requirement in the database.

**Comment by nurul3101** [comment]
Hello @ErikDakoda 👋 

At the moment, this is a requirement to have relations defined on both sides of the models. 

We have a GitHub Issue discussing how we can improve this:

- https://github.com/prisma/prisma/issues/2018

Can you have a look at the GitHub Issue? The issue also discusses the original reasoning behind the motivation to make opposite relations required ([Reference Comment](https://github.com/prisma/prisma/issues/2018#issuecomment-1409239477)).

**Reply by ErikDakoda** [reply]
@nurul3101, thank you for your answer. I have read through all relevant threads, including the one you mentioned, which has been going on for three and a half years and seems to be going nowhere. The explanation "The original reasoning was that if there is a relation from A to B, there always also is a way to go from B to A in the database" is not really a technical explanation of why this is required. In fact "in the database" B has no references to A - this is a virtual field added by Prisma in code only.

Here is the problem: Almost every model references User, which then has to reference every model back, and this creates very tightly coupled code and bad system architecture. It is impossible to break your code into packages because this would lead to circular dependencies.

I know that Prisma also has the odd limitation of having to define your entire schema in a single file, but a wonderful package called ZenStack removes this limitation. This allows more readable and better-organized code. But because of the pointless reverse reference rule, your code is still tightly coupled and can't be broken into packages and can't be made composable.

I am puzzled by Prisma's resistance to changing this, even after all the examples users have given in the mentioned threads where a one-way relation would be logical and desirable. I have not seen any technical explanation of why this would be a difficult change.

**Reply by janpio** [reply]
I don't disagree that it would be nice if Prisma worked different here. But it does not, and changing it is not that simple.

> "The original reasoning was that if there is a relation from A to B, there always also is a way to go from B to A in the database" is not really a technical explanation of why this is required. In fact "in the database" B has no references to A - this is a virtual field added by Prisma in code only.

If there is a foreign key from A to B, you can always also query for all Bs that are referenced in A.

> I am puzzled by Prisma's resistance to changing this, even after all the examples users have given in the mentioned threads where a one-way relation would be logical and desirable. I have not seen any technical explanation of why this would be a difficult change.

It would be a breaking change to all existing Prisma users, as then for each relation you have to make a choice: Do I want to have a relation in both direction, or only in one? All our tooling would suddenly consider both cases. (If you think this is trivial to figure out and implement, I am looking forward to your PR. If it is sound, we would be happy to resolve this problem.)

Overall it is not a big deal for most. If you don't want to use the back relation, just don't use it.

**Reply by ErikDakoda** [reply]
I agree, if this would be a breaking change, it would not be worth it. I would instead implement it as a new attribute such as `@oneway` or `@pointer` or a new argument such as `@relation(fields: [authorId], references: [id], oneway: true)`.  I would have to really familiarize myself with the Prisma repo before I would attempt a PR for such a change.

**Reply by janpio** [reply]
Yes, that is also my current expectation - we would need to configure it some way. I have not found any really elegant way to do so :/

**Reply by metafab** [reply]
Instead of a new attribute, maybe having @ignore work with relations would partly solve the issue?
See #21244

**Reply by avilabiel** [reply]
Guys, I was expecting an @ignore to get rid of this. Not having an @ignore really surprises me.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
